### ISSUE TITLE
Support pin-depends

### ISSUE THREAD

**Issue by pascutto** [ocurrent/ocaml-ci#101]
Say I have two packages:
- `A`
- `B`, which depends on the `master` branch of `A` in `opam`'s `pin-depends` field.

Now I have the following scenario:
- Commit to `B`, CI runs normally.
- Commit to `A`, *adding a depext `D`*.
- Commit to `B`.
When the CI runs for that last commit, it fails during `A` install because of missing system dependency (it did not install `D`).

It seems that the pin is cached by the CI during the `depext` stage, and can't detect upstream changes, since the pin is against a branch name instead of a commit, and the pinned package is not pulled at this moment.

/cc @Julow

**Comment by Julow** [comment]
We tried to study a simple/naive fix yesterday:
We could use `git ls-remote` to query the last rev of each pin-depends (in a job placed outside of `Analyse`'s cache). The rev could then be inserted in the command installing them, invalidating Docker's cache if that changes.
The problem is that opam files and `opam show` are very hard to parse. To solve that we should use Opam's `opam-format` library.

**Comment by talex5** [comment]
Note: another way to depend on dev versions is to use `git submodule` to add the exact commit you want to your project. ocaml-ci will automatically fetch these submodules (using `git clone --recursive`) and dune will automatically build against them.

**Comment by pascutto** [comment]
> Note: another way to depend on dev versions is to use `git submodule` to add the exact commit you want to your project. ocaml-ci will automatically fetch these submodules (using `git clone --recursive`) and dune will automatically build against them.

If we wanted to depend on a specific commit we could also pin-depend on that specific commit, and the CI works well in that case too.

**Comment by talex5** [comment]
Note that `pin-depends` is broken at the moment as the new solver (in #167) only has access to the Git clone and opam-repository. I guess we'll need to implement fetching of pin-depends manually somehow.

**Comment by CraigFe** [comment]
Title changed to reflect that this is now a more general issue.

**Comment by talex5** [comment]
pin-depends with a commit hash is working again now. I don't think we'll support pinning to branches any time soon.

**Comment by pascutto** [comment]
It seems like there is an issue when pinning to a repo where the opam file is not at the root (maybe it's unrelated to pinning?), it can't find the opam file thus fails during analysis.

**Comment by talex5** [comment]
Do you have a link to an example? ocaml-ci currently assumes the opam file is at `/$name.opam`, although I think opam also allows some other forms.

**Comment by pascutto** [comment]
Here's one: https://ci.ocamllabs.io/github/vocal-project/vocal/commit/363a91d49e1e89c202d5656c088f3a8ca538f68c/variant/(analysis)
The opam file is under a `opam/` folder

**Comment by talex5** [comment]
I've pushed a fix - should be working now!
